File: DataEngineering/03_Data_Ingestion/projects/golang/04_webhook_listener/main.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/google/uuid"
)

type WebhookHandler struct {
	S3Client *s3.Client
	Bucket   string
}

func (h *WebhookHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "Only POST allowed", http.StatusMethodNotAllowed)
		return
	}

	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, "Failed to read body", http.StatusInternalServerError)
		return
	}
	defer r.Body.Close()

	if len(body) == 0 {
		http.Error(w, "Empty payload", http.StatusBadRequest)
		return
	}

	key := fmt.Sprintf("webhooks/%s_%s.json", time.Now().Format("20060102150405"), uuid.New().String())

	_, err = h.S3Client.PutObject(r.Context(), &s3.PutObjectInput{
		Bucket:      aws.String(h.Bucket),
		Key:         aws.String(key),
		Body:        bytes.NewReader(body),
		ContentType: aws.String("application/json"),
	})

	if err != nil {
		log.Printf("Failed to upload to S3: %v", err)
		http.Error(w, "Internal Server Error", http.StatusInternalServerError)
		return
	}

	log.Printf("Received webhook, saved to s3://%s/%s", h.Bucket, key)
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	w.Write([]byte(`{"status": "success"}`))
}

func main() {
	log.Println("Starting Data Engineering Project: 04_webhook_listener (AWS S3 Backend)")

	bucket := os.Getenv("S3_BUCKET")
	if bucket == "" {
		bucket = "my-webhook-payloads"
	}

	cfg, err := config.LoadDefaultConfig(context.TODO())
	if err != nil {
		log.Fatalf("unable to load SDK config, %v", err)
	}

	s3Client := s3.NewFromConfig(cfg)

	handler := &WebhookHandler{
		S3Client: s3Client,
		Bucket:   bucket,
	}

	http.Handle("/webhook", handler)

	port := os.Getenv("PORT")
	if port == "" {
		port = "8080"
	}

	log.Printf("Listening for webhooks on port %s...", port)
	if err := http.ListenAndServe(":"+port, nil); err != nil {
		log.Fatal(err)
	}
}

File: DataEngineering/03_Data_Ingestion/projects/python/04_webhook_listener/main.py
import os
import json
import uuid
import logging
from datetime import datetime
from fastapi import FastAPI, Request, HTTPException
import uvicorn
import boto3
from botocore.exceptions import BotoCoreError, ClientError

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

app = FastAPI(title="Webhook Listener API")

S3_BUCKET = os.getenv("S3_BUCKET", "my-webhook-payloads")
s3_client = boto3.client('s3', region_name=os.getenv("AWS_REGION", "us-east-1"))

@app.post("/webhook")
async def receive_webhook(request: Request):
    """
    Endpoint to receive webhook payloads and upload them directly to AWS S3.
    """
    try:
        payload = await request.json()
    except json.JSONDecodeError:
        raise HTTPException(status_code=400, detail="Invalid JSON payload")

    # Generate a unique object key based on timestamp and uuid
    timestamp = datetime.utcnow().strftime("%Y%m%d%H%M%S")
    file_id = str(uuid.uuid4())
    object_key = f"webhooks/{timestamp}_{file_id}.json"
    
    try:
        # Upload the payload to S3
        s3_client.put_object(
            Bucket=S3_BUCKET,
            Key=object_key,
            Body=json.dumps(payload),
            ContentType="application/json"
        )
        logging.info(f"Successfully uploaded payload to s3://{S3_BUCKET}/{object_key}")
    except (BotoCoreError, ClientError) as e:
        logging.error(f"Failed to upload to S3: {e}")
        # In a real scenario, you might write to a local fallback or Kafka dead-letter queue
        raise HTTPException(status_code=500, detail="Failed to persist webhook payload")

    return {"status": "success", "key": object_key}

def main():
    logging.info('Starting Data Engineering Project: 04_webhook_listener (FastAPI + Boto3)')
    port = int(os.getenv("PORT", 8080))
    # Uvicorn is a standard ASGI server for FastAPI
    uvicorn.run(app, host="0.0.0.0", port=port, log_level="info")

if __name__ == '__main__':
    main()
